src/NativeScript/ObjC/Constructor/ObjCConstructorBase.mm

```
@@ -146,8 +148,95 @@ static void writeAdapter(ExecState* execState, const JSValue& value, void* buffe
+    }
+
+    return builder.toString();
+}
+
+static JSValue getInitializerForSwiftStyleConstruction(ExecState* execState, JSValue prototype, WTF::String jsName) {
+    static WTF::String initWith("initWith", WTF::String::ConstructFromLiteral);
+    static WTF::String init("init", WTF::String::ConstructFromLiteral);
+    static WTF::String error("Error", WTF::String::ConstructFromLiteral);
+
+    JSValue value;
+#define RETURN_IF_CELL if (value.isCell()) return value
+    
+    if (jsName.isEmpty()) {
+        return value;
+    }
+
+    value = prototype.get(execState, Identifier::fromString(execState, WTF::String(initWith + jsName)));
+    RETURN_IF_CELL;
+
+    value = prototype.get(execState, Identifier::fromString(execState, WTF::String(initWith + jsName + error)));
+    RETURN_IF_CELL;
+
+    value = prototype.get(execState, Identifier::fromString(execState, WTF::String(init + jsName)));
+    RETURN_IF_CELL;
+
+    value = prototype.get(execState, Identifier::fromString(execState, WTF::String(init + jsName + error)));
+    RETURN_IF_CELL;
+
+    return JSValue();
+}
+
 static EncodedJSValue JSC_HOST_CALL constructObjCClass(ExecState* execState) {
     ObjCConstructorBase* constructor = jsCast<ObjCConstructorBase*>(execState->callee());
+
+    if (execState->argumentCount() <= 1) {
+        MarkedArgumentBuffer initializerArguments;
+        JSValue initializer;
+        if (JSFinalObject* argument = jsDynamicCast<JSFinalObject*>(execState->argument(0))) {
+            initializer = getInitializerForSwiftStyleConstruction(execState, constructor->instancesStructure()->storedPrototype(), computeInitializerJSName(execState, argument, initializerArguments));
+        } else if (execState->argumentCount() == 0) {
+            initializer = constructor->instancesStructure()->storedPrototype().get(execState, Identifier::fromString(execState, "init"));
+        } else {
+            goto new_legacy;
+        }
+
+        if (initializer && initializer.isCell()) {
```

If `getInitializerForSwiftStyleConstruction` doesn't find a suitable initializer corresponding to the given literal, do we want to fallback to the legacy `new` resolution or maybe throw an error (e.g. 'no initializer matching the given signature').
Is there any case in which `new` will be called with a literal (`FinalObject`) without targeting the new swift-style initializer resolution?